source/MaterialXGenShader/ColorManagementSystem.cpp

```diff
@@ -0,0 +1,94 @@
+#include <MaterialXGenShader/ColorManagementSystem.h>
+
+#include <MaterialXGenShader/ShaderGenerator.h>
+#include <MaterialXGenShader/Nodes/SourceCodeNode.h>
+
+namespace MaterialX
+{
+
+ColorSpaceTransform::ColorSpaceTransform(const string& ss, const string& ts, const TypeDesc* t)
+    : sourceSpace(ss)
+    , targetSpace(ts)
+    , type(t)
+{
+    if (type != Type::COLOR3 && type != Type::COLOR4)
+    {
+        throw ExceptionShaderGenError("Color space transform can only be a color3 or color4.");
+    }
+}
+
+
+ColorManagementSystem::ColorManagementSystem(ShaderGenerator& shadergen, const string& configFile)
+    : _configFile(configFile)
+    , _shadergen(shadergen)
+{
+}
+
+bool ColorManagementSystem::registerImplementation(const ColorSpaceTransform& transform, CreatorFunction<ShaderNodeImpl> creator)
+{
+    if (transform.type)
+    {
+        string implName = getImplementationName(transform);
+        _implFactory.registerClass(implName, creator);
+        _registeredImplNames.push_back(implName);
+        return true;
+    }
+    return false;
+}
+
+void  ColorManagementSystem::setConfigFile(const string& configFile)
+{
+    _configFile = configFile;
+    _implFactory.unregisterClasses(_registeredImplNames);
+    _registeredImplNames.clear();
+    _cachedImpls.clear();
+}
+
+void ColorManagementSystem::loadLibrary(DocumentPtr document)
+{
+    _document = document;
+    _implFactory.unregisterClasses(_registeredImplNames);
+    _registeredImplNames.clear();
+    _cachedImpls.clear();
+}
+
+ShaderNodePtr ColorManagementSystem::createNode(const ColorSpaceTransform& transform, const string& name)
+{
+    string implName = getImplementationName(transform);
+    ImplementationPtr impl = _document->getImplementation(implName);
+    if (!impl)
+    {
+        throw ExceptionShaderGenError("No implementation found for transform: ('" + transform.sourceSpace + "', '" + transform.targetSpace + "').");
+    }
+
+    // Check if the shader implementation has been created already
+    ShaderNodeImplPtr shaderImpl;
+    auto it = _cachedImpls.find(transform);
+    if (it != _cachedImpls.end())
+    {
+        shaderImpl = it->second;
+    }
+    // If not, try creating a new shader implementation in the factory
+    else
+    {
+        shaderImpl = _implFactory.create(implName);
+    }
+    // Fall back to the default implementation
+    if (!shaderImpl)
+    {
+        shaderImpl = SourceCodeNode::create();
+    }
+    shaderImpl->initialize(impl, _shadergen);
+
+    _cachedImpls[transform] = shaderImpl;
+    ShaderNodePtr shaderNode = ShaderNode::createColorTransformNode(getShaderNodeName(transform, name), shaderImpl, transform.type, _shadergen);
+
+    return shaderNode;
+}
+
+string ColorManagementSystem::getShaderNodeName(const ColorSpaceTransform& transform, const string& name)
+{
+    return name + "_" + transform.sourceSpace + "_to_" + transform.targetSpace + "_" + transform.type->getName();
```

Comment:
Now I'm gonna rant about the node naming again ;) 

Keeping in mind that the node name is used for the _variable_  storing the result in generated code I'm not sure we need to do these long string concatenations to make the code readable. The _function_ calculating the transform will still be named according to the color spaces it is transforming, the variable storing the result doesn't need this information. To my mind it's enough if the variable says which node it came from and that it stores a color transformed value. So "nodename_cm" would be enough. For the case of multioutput we need "nodename_output_cm". The client side can handle this, since there you know the origin of the color space transform.

One can also argue that readability of the generated code should not be more important than codegen speed or code size, since in general this code is not for humans to read anyway. *end of rant*

Reply:
I guess I don't feel this is much of a performance issue and I wouldn't think the code size would have any real impact. I think the readability is nice. But I don't feel strongly enough to argue about it, so I'll make the change.